pymtl3/passes/backends/sverilog/import_/ImportPass.py

```diff
@@ -831,122 +694,169 @@ def _get_arg_str( name, obj ):
+
+      if isinstance( dtype, rdt.Struct ):
+        # We don't create a new struct if we are copying values from pymtl
+        # land to verilator, i.e. this port is the input to the imported
+        # component.
+        dtype_name = dtype.get_class().__name__
+        if dtype_name not in symbols:
+          symbols[dtype_name] = dtype.get_class()
+
+        # We create a long Bits object tmp first
+        ret = [ f'tmp = Bits{dtype_nbits}(0)' ]
+
+        # Then we write each struct field to tmp
+        body, pos = s._gen_struct_write( 'i', rhs, 'tmp', dtype, 0 )
+        ret.extend(body)
+
+        # At the end, we write tmp to the corresponding CFFI variable
+        ret.extend( s._gen_ref_write( lhs, 'tmp', dtype_nbits ) )
+
+        assert pos == dtype_nbits
+        return ret
+
+      assert False, f"unrecognized data type {dtype}!"
+
     else:
       ret = []
       for idx in range( n_dim[0] ):
         _lhs = f"{lhs}[{idx}]"
         _rhs = f"{rhs}[{idx}]"
-        ret += s.gen_port_array_input( _lhs, _rhs, dtype, n_dim[1:] )
+        ret.extend( s.gen_port_array_input( _lhs, _rhs, dtype, n_dim[1:], symbols ) )
       return ret
 
-  def gen_comb_input( s, packed_ports ):
+  def gen_comb_input( s, packed_ports, symbols ):
     ret = []
     # Read all input ports ( except for 'clk' ) from component ports into
     # the verilated model. We do NOT want `clk` signal to be read into
     # the verilated model because only the sequential update block of
     # the imported component should manipulate it.
-    for py_name, v_name, rtype in packed_ports:
+
+    for pname, vname, rtype in packed_ports:
       p_n_dim, p_rtype = s._get_rtype( rtype )
-      if s._get_direction( p_rtype ) == 'InPort' and py_name != 'clk' and v_name:
+      if s._get_direction( p_rtype ) == 'InPort' and pname != 'clk' and vname:
         dtype = p_rtype.get_dtype()
-        lhs = "_ffi_m."+s._verilator_name(v_name)
-        rhs = "s.mangled__"+s._verilator_name(py_name)
-        ret += s.gen_port_array_input( lhs, rhs, dtype, p_n_dim )
+        lhs = "_ffi_m."+s._verilator_name(vname)
+        rhs = f"s.{pname}"
+        ret += s.gen_port_array_input( lhs, rhs, dtype, p_n_dim, symbols )
     return ret
 
   #-------------------------------------------------------------------------
   # gen_comb_output
   #-------------------------------------------------------------------------
 
-  def gen_port_array_output( s, lhs, rhs, dtype, n_dim ):
-    nbits = dtype.get_length()
+  def gen_port_array_output( s, lhs, rhs, dtype, n_dim, symbols ):
     if not n_dim:
-      return s._gen_ref_read( lhs, rhs, nbits )
+      dtype_nbits = dtype.get_length()
+
+      # If the top-level signal is a Bits, we directly
+
+      if isinstance( dtype, rdt.Vector ):
+        return s._gen_ref_read( lhs, rhs, dtype_nbits )
+
+      # If the top-level signal is a struct, we add the datatype to symbol?
+
+      if isinstance( dtype, rdt.Struct ):
```

Comment:
If `dtype` is a nested struct, does this work? I would imagine there would be a recursive process to breakdown the data type into bits before passing into `_gen_ref_read`?

Reply:
hmm good point. Can you add a test to catch that?

Reply:
This doesn't matter, because tmp is going to hold all the value. Line 835 will recursively break tmp down to the actualy nested struct. See line 836 where the final position is dtype,get_length.